### ISSUE TITLE
VS 2019 16.7.0 Prev 2.0 - Can't deploy in one step Debug or Release

### ISSUE THREAD

**Issue by gmck** [xamarin/xamarin-android#4780]
Can no longer Build and Deploy in one step. The build will complete successfully and then the deployment will start and end immediately with the message Deploy failed - no other info. I reported something quite similar a few builds ago and it was fixed - so it appears to have come back again. The only option left at that point is to clean, build and deploy in separate steps and all is good again, until the next edit.

Earlier today I uploaded a zip that demonstrates this behaviour to GooglePlayServicesCompopnents. Now having gone back to my app and included the modified code from that app, my app also does the same thing. 

Also still can't open any layout files in this project without the error reported in issue #4317 (Feb 25). Has no effect at runtime, but it would be nice to have it fixed and know why this is happening.

**Comment by gmck** [comment]
Forgot the following

Re the can't build and deploy in one step, I occasionally see the following

Error		Your project file doesn't list 'win' as a "RuntimeIdentifier". You should add 'win' to the "RuntimeIdentifiers" property in your project file and then re-run NuGet restore.

I've never seen that before in VS and have no idea what it means. As per above it goes away after a clean or deletes of obj/bin etc.

**Comment by dellis1972** [comment]
@gmck Thanks for the bug report. We need some diagnostic build logs in order to investigate this further. Specifically we need to Deployment log. 

These docs should help you get that information https://docs.microsoft.com/en-gb/xamarin/android/troubleshooting/troubleshooting#Diagnostic_MSBuild_Output. And you should be able to attach it to this issue

**Comment by gmck** [comment]
@dellis1972  As stated above, I've already uploaded the project. Is that good enough or do you still require the deployment log?

**Comment by dellis1972** [comment]
@gmck did you mean https://github.com/xamarin/GooglePlayServicesComponents/issues/350 ? Because there doesn't seem to be an attachment on that either.

**Comment by brendanzagaeski** [comment]
Oh ha, I just saw this conversation by chance. I think what happened was moljac edited [this comment](https://github.com/xamarin/GooglePlayServicesComponents/issues/350#issuecomment-640134797). If you open the edit history drop-down on that comment, the link is there.

**Comment by dellis1972** [comment]
@gmck I got the sample you sent via email. I am unable to replicate the deployment problems in VS 2019 Preview. I'll keep trying. 

The ZipException you mentioned in your email, that is a bit weird. but we will need a full diagnostic log in order to figure the one out. Again I can't repo that with the sample you sent.

**Comment by dellis1972** [comment]
@gmck I might be easier to just install https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools. This is an excellent extension which will give you access to all the logs for the build VS does. It makes it very easy to capture the binlog files which will help us diagnose these kinds of issue

**Comment by gmck** [comment]
@dellis1972  replied via email.

**Comment by gmck** [comment]
@dellis1972. Dean attached is an example set of logs that demonstrate the failure of the 1 step deploy.

All I did was deploy to a different phone. Previously I was deploying to a Pixel3a and it was building and deploying ok and then I decided to swap to my own phone to test the latest changes.

Hit deploy and it built and then failed to deploy. I can't see in red icons in the logs, but the buildout states build succeeded, deploy failed and the apk is not on the phone. The Build Message list is empty.

Included in the logs.zip is the output from the Output window. This is the typical behaviour I've been seeing with this app and the GoogleSignIn project you already have.

You fix it by Clean, Build and Deploy all separate steps, until the next time. Note that this a debug build not a Release build as per your comments in your last email.

[logs.zip](https://github.com/xamarin/xamarin-android/files/4790037/logs.zip)

**Comment by gmck** [comment]
@dellis1972 Another example. All I did was change snackbar.SetDuration() in a ShowSnackBar() method of the MainActivity from 1000 to 500.  Hit Deploy - and got

Done executing task "ResolveNuGetPackageAssets" -- FAILED.
Done building target "ResolveNuGetPackageAssets" in project "OBDNowPro.csproj" -- FAILED.
Done building project "OBDNowPro.csproj" -- FAILED.
Build FAILED.
2>Deployment failed to Samsung SM-G981B.
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Deploy: 0 succeeded, 1 failed, 0 skipped ==========

Attaching the logs
[logs1.zip](https://github.com/xamarin/xamarin-android/files/4790383/logs1.zip)

**Comment by dellis1972** [comment]
@gmck can you try adding the following to your csproj

```
    <BaseNuGetRuntimeIdentifier>monoandroid</BaseNuGetRuntimeIdentifier>
```

In the top most PropertyGroup should be fine. That might help with the Nuget issue.

**Comment by gmck** [comment]
Started from where I left off last night. Deleted the apk from the phone. Added <BaseNuGet.. etc and hit Deploy. Starting building etc and then VS crashed and restarted and of course, then there was nothing in the Build logging tab. I found the log in the temp folder and opened it with MS Build Structured Log Viewer (is there a way to open a log from within VS - when there are no files in  Build logging tab)? The Open logs menu item doesn't open any dialog if there are no files.

Opening the log didn't reveal anything that looked like an error. I'll include the log here.

Went back and did a Clean, Build and Deploy and everything was back to normal.

[OBDNowPro_Release_AnyCPU_Build_2020-06-18T08_50_27.2573123+10_00.zip](https://github.com/xamarin/xamarin-android/files/4795510/OBDNowPro_Release_AnyCPU_Build_2020-06-18T08_50_27.2573123%2B10_00.zip)

**Comment by gmck** [comment]
@dellis1972 Needed to make an aab, so I tried Deploy one step which built but failed to deploy. This log at least shows an error. Also included the output from build.

[OBDNowPro_Release_AnyCPU_Build_2020-06-18T10_29_40.0002258+10_00.zip](https://github.com/xamarin/xamarin-android/files/4795799/OBDNowPro_Release_AnyCPU_Build_2020-06-18T10_29_40.0002258%2B10_00.zip)

**Comment by dellis1972** [comment]
This is very weird. The Logs just stop at this point in the build.

```
arget "_SetLatestTargetFrameworkVersion: (TargetId:8)" in file "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Xamarin\Android\Xamarin.Android.Tooling.targets" from project "C:\ProjectsVS2019Preview\OBDNowPro\OBDNowPro\OBDNowPro.csproj" (target "_CreatePropertiesCache" depends on it):
Set Property: TargetFrameworkMoniker=MonoAndroid,Version=v10.0
Set Property: NuGetTargetMoniker=MonoAndroid,Version=v10.0
Set Property: TargetFrameworkMonikerAssemblyAttributesPath=obj\Release\100\MonoAndroid,Version=v10.0.AssemblyAttributes.cs
Added Item(s): FileWrites=obj\Release\100\MonoAndroid,Version=v10.0.AssemblyAttributes.cs
Added Item(s): 
    Reference=
        Java.Interop
        System.Runtime
Done building target "_SetLatestTargetFrameworkVersion" in project "OBDNowPro.csproj".: (TargetId:8)
Target "_SetLatestTargetFrameworkVersion" skipped. Previously built successfully.
Target "_SetLatestTargetFrameworkVersion" skipped. Previously built successfully.
Target "GetFrameworkPaths: (TargetId:9)" in file "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets" from project "C:\ProjectsVS2019Preview\OBDNowPro\OBDNowPro\OBDNowPro.csproj" (target "GetReferenceAssemblyPaths" depends on it):
Done building target "GetFrameworkPaths" in project "OBDNowPro.csproj".: (TargetId:9)
Target "GetReferenceAssemblyPaths: (TargetId:10)" in file "C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\Microsoft.Common.CurrentVersion.targets" from project "C:\ProjectsVS2019Preview\OBDNowPro\OBDNowPro\OBDNowPro.csproj" (target "_ResolveMonoAndroidSdks" depends on it):
Set Property: TargetFrameworkDirectory=
Task "GetReferenceAssemblyPaths" (TaskId:8)
```

No errors or warnings. Its very very odd. @gmck could you email me your csproj for this project? I'd like to see if I can look at the settings for it to see if there is anything that might be causing this.

With regards to opening a Log in the Structured Viewer, you should be able to right click on the entry in the Build Logging tab in Visual Studio and see an "Open Logs External" menu option. You should also be able to Save the log to a more user friendly folder from that right click menu as well.

**Comment by dellis1972** [comment]
@jonathanpeppers Could use some of your Windows insight on this if you don't mind. There appears to be something very weird going on.

**Comment by jonathanpeppers** [comment]
@gmck when you recorded logs with Project System Tools, can you share a screenshot of every build that ran?

I'm wondering if some other build in the IDE was started that canceled the build running `Install`.

Also, does VS 2019 16.6 stable have the same issue? It seems like you are using 16.7 preview, so maybe there is an issue with preview at the moment.

**Comment by gmck** [comment]
@jonathanpeppers attaching screenshot. Result build succeeded, deploy fail.

![Capture](https://user-images.githubusercontent.com/6430011/85076583-bf28f580-b203-11ea-8792-06fe72b0e7ff.PNG)

**Comment by gmck** [comment]
@dellis1972 @dellis1972 . Came across something in the csprog that I didn't mean to have in there that may be effecting this although I don't understand why.

I usually keep a docs folder where I just keep a todo list sort of thing, the sort of comments that are too big to keep in the code. Usually, I just exclude that folder from the solution. But I noticed just now that it was included. The file was last updated in late May, but I can't say if it has been included since that time, but since it hasn't been changed then there is a fair chance that it has been included all that time.

I just excluded it and was able a step Deploy successfully for the first time. Here is how it was.
 
```
<ItemGroup>
    <ProguardConfiguration Include="proguard.cfg" />
    <GoogleServicesJson Include="google-services.json" />
    <None Include="Docs\DialogFraments" />
    <None Include="Resources\AboutResources.txt" />
    <None Include="Properties\AndroidManifest.xml" />
    <None Include="Assets\AboutAssets.txt" />
    <AndroidResource Include="Resources\drawable\bmw5.webp" />
  </ItemGroup>

```
I would have presumed that the None Include would not have had adverse effects on the build, but maybe it does.

@dellis1972 . The GoogleSignIn app for Mel also had the "Can't deploy in 1 step" problem, so maybe that is worth another look or pass it on the Jonathon.

@jonathanpeppers I'll try this in 16.6 stable and check how it goes.

**Comment by gmck** [comment]
@jonathanpeppers Upgraded to VS 16.6.2 and tried the GoogleSignInApp. 1 step deploy doesn't work there either.  Have previously emailed that project to @dellis1972 if you want to try. Can't upload it here, see comments earlier. Dean has previously tried to repo it with no success.

Got the following.
```
Removing manifest/AndroidManifest.xml as it is not longer required.
XABBA7000: Xamarin.Tools.Zip.ZipException: Invalid argument
   at Xamarin.Tools.Zip.ZipArchive.ReadEntry(UInt64 index)
   at Xamarin.Android.Tasks.BuildBaseAppBundle.FixupArchive(ZipArchiveEx zip)
   at Xamarin.Android.Tasks.BuildApk.ExecuteWithAbi(String[] supportedAbis, String apkInputPath, String apkOutputPath)
   at Xamarin.Android.Tasks.BuildApk.RunTask()
   at Xamarin.Android.Tasks.AndroidTask.Execute()
Output Item(s): BaseZipFile=
Done executing task "BuildBaseAppBundle" -- FAILED.
Done building target "_BuildApkEmbed" in project "GoogleSignInTest.csproj" -- FAILED.
Done building project "GoogleSignInTest.csproj" -- FAILED.
Build FAILED.
2>Deployment failed to Google Pixel 3a.
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Deploy: 0 succeeded, 1 failed, 0 skipped ==========

```
3 step - clean, build, deploy always works. Change a couple of lines, comment or uncomment the same lines and 1 step deploy fails both debug and release builds - so basically no different to  Preview 16.7.0 Preview 2.0

I will try my main project next with 16.6.2.

**Comment by gmck** [comment]
@jonathanpeppers. Built my main project with 16.6.2. Added quite a bit of code over a couple of hours without any problems. Just hit deploy each time and then checking the results with app running. So 16.6.2 appears to work reliably in debug mode doing incremental builds.

Swapped to Release and it 1 step Deployed ok but crashed with a class not found exception at startup. That was a proguard problem had to add the following to proguard. cfg

`-keep class com.google.android.material.internal.BaselineLayout`

The next Deploy failed, but that was because I also left an uncommented line in cfg file. Fixed that and 1 step deploy worked fine and the app ran fine. I have one request, could you propagate the error re the uncommented line error etc and report it at the end of the build. Takes a lot of searching of the build output using "r8" to actually find what made it barf. The info was there, but it would be a lot nicer if you didn't have to search for it. The build logging files aren't much help when looking for the cause of an error like that.

I then swapped back to the Preview 16.7.0 P2.0 and the 1 step Deploy worked for the release build- which was pretty amazing!!

I'll go back to the debug release with the Preview and keep adding code and see how it goes incrementally.

**Comment by gmck** [comment]
@dellis1972 , @jonathanpeppers 

Back working with the debug build of my main app. 1 step deploy is still failing. I made a really small change to strings.xml and tried a 1 step deploy. It failed with the usual lines at end of build output

```
2>Starting deployment to Google Pixel 3a ...
2>Deploying to Google Pixel 3a ...
Build started.
Project "OBDNowPro.csproj" (Install target(s)):
2>Deployment failed to Google Pixel 3a.
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
========== Deploy: 0 succeeded, 1 failed, 0 skipped ==========
```
I'm also attaching a screenshot of the Build logging tab. It indicates the install log is still running, when obviously it is not. 

![BuildLoggingtab](https://user-images.githubusercontent.com/6430011/85188871-36868400-b2ed-11ea-8918-676543415599.PNG)

Opening the install log shows a Title Exceptions, expanding it reveals 
Value cannot be null.
Parameter name: source
As shown in the next screenshot.  

![InstallBuildLog](https://user-images.githubusercontent.com/6430011/85188991-363ab880-b2ee-11ea-9034-9e871331626c.PNG)

**Comment by gmck** [comment]
@dellis1972, @jonathanpeppers 
Some more observations re 1 step deploy failures. I'm finding that after adding or modifying code in some cases when the deploy fails, I can just hit Deploy again (without having to clean, build, deploy) and the deploy then succeeds.

I'm attaching a screenshot of when it fails. Which shows the Install target still running and then two entries for GetExtraLibaryLocationsForDesigner and PrepareResources:_GenerateCompileInputs finished. Why the two entries?
![BuildLogTab_AfterFailedDeploy](https://user-images.githubusercontent.com/6430011/85214363-e111ac00-b3ac-11ea-9894-3f21aadde3f4.PNG)

The second screenshot shows the result of a successful Deploy after the first one failed.  It still has those two extra entries appearing after the install succeeds, but why is the timestamp always later than the install timestamp and is that significant? I would have thought that install would be the last target to run and did not expect to see anything appearing or finishing after install.

![BuildLogTab_AfterSuccessfulBuildThenDeploy](https://user-images.githubusercontent.com/6430011/85214392-1a4a1c00-b3ad-11ea-9d07-5af44a530dee.PNG)

**Comment by jonathanpeppers** [comment]
@gmck when you are modifying code and hit this issue. Are you changing Android layout `.axml` files? I think the Android designer might be involved here.

If you use the plain XML editor (`Open With... | XML (Text Editor)`, does the problem go away?

**Comment by gmck** [comment]
@jonathanpeppers I don't have any .axml files in the project. All my layouts are .xml files. My old project used axml (and they all open without errors). When I started to convert the app to Android 10 I converted all to xml mainly because many of the files had to be converted to androidx equivalents. I redesigned many of them in Android Studio especially the androidx.constraint layouts and then created new layout files in VS and copied/pasted from AS to VS. I may have even imported some use Add existing item...

None of these files will open without error in VS. See issue #4317. It has been the same with every preview since. None of them gives any problem re errors when building and they all run on the device(s) as designed. Since I reported this problem I've only added 2 xml layouts.

They all open fine with Notepad or even VS Preview if you use Open With... it is just the designer that gives the error.

In contrast, the GoogleSignIn app (xml layouts) I uploaded previously does open its xml files without the error.

Does any of that help?
